### DISCUSSION TITLE
ESP32-C6 Support?

### DISCUSSION THREAD

**Discussion by Jibun-no-Kage** [micropython/micropython#11356] [ESP32]
I know I am pushing the curve here, but MP support for C6?

"C:\Program Files (x86)\Thonny\python.exe" -u -m esptool --port COM13 erase_flash
esptool.py v3.1
Serial port COM13
Connecting....
Detecting chip type...

A fatal error occurred: Unexpected CHIP magic value 0x2ce0806f. Failed to autodetect chip type.

Erase command returned with error code 2

**Comment by scruss** [comment]
That's not MicroPython's problem that you're seeing. I don't think esptool v3.1 knows about esp32c6. v4.4 does, though.

Matt Trentini was talking about the C6 at the Melbourne meetup yesterday. I think he may have managed a build, but there was some detail about the Espressif build system version that made it difficult. For me, anything that Matt describes as "difficult" is pretty much nope-out territory

**Comment by Jibun-no-Kage** [comment]
This is not a problem.... it is a question.  ESP32-C6 is not supported by anything right now, including, Arduino IDE, Circuit Python, etc.  So the question is, again, when will MP support ESP32-C6?  So... what you are saying is... there is no established timeline or target date as yet?

**Reply by scruss** [reply]
Matt likely has a more helpful answer, but there are never any established timelines or target dates here.

ESP32-C6 apparently requires Espressif IDF 5.1, which none of Arduino / CircuitPython / etc use yet. This came from one of the CircuitPython devs on the MicroPython Discord

**Reply by scruss** [reply]
> This is not a problem.... 

Having an old version of esptool that doesn't support `--chip esp32c6` will very much be a problem for *you*, though

**Comment by Jibun-no-Kage** [comment]
Yes, 5.1 is still pending.  The official documentation that comes with the ESP32-C6 boards states this.  And that is the key reason why Arduino IDE is going to be pending for a while since it always lags official IDE support.  And @scruss, you nailed the reason for my question... the hope that Matt has some inside track, or at least one can hope, he does.  Last, if there was something official, this discussion would not exist, right?  But as I said in my initial question... I know I am pushing the curve here.  :)

**Comment by Jibun-no-Kage** [comment]
Just was illustrative of the chip ID.  That it really was a ESP32-C6.  I knew already I could not flash with that version.

**Comment by andypiper** [comment]
I have a couple of the 01Space ESP32-C6 boards, as does @mattytrentini - I think the conversation at the meetup on Wednesday was indeed that we need to wait for IDF 5.1 before we can even attempt to compile for it. Rest assured that a few folks in the community are interested and will see what we can do, when we can; but no, there's no immediate news on when this might be available.

**Comment by Jibun-no-Kage** [comment]
I can't say I don't agree, that seems logical, reasonable.  If it is a bit of a bummer.  These C6 boards are really interesting, so looking forward to when MP supports them.  Oh, I got mine from Adafruit, so they are DevKitC-1s.

**Comment by mattytrentini** [comment]
So...as you've both alluded to, the main difficulty is in building MicroPython with the 5.x toolchain - not so much C6 support specifically. Different compiler, many more warnings/errors - and a lot are in libraries we're _using_ (Bluetooth looked particularly difficult). 

I think once we're able to successfully build for _any_ ESP32 on 5.x then C6 support won't take long.

BTW, Espressif conveniently publishes their toolchains as Docker containers, so it's pretty straightforward to try out their [5.1 release branch](https://hub.docker.com/layers/espressif/idf/release-v5.1/images/sha256-b9a795449dd46ae264e07c19aa639e0542dacb36fee0b1611af32eafda46261e?context=explore) for example. 

Investigating an update to newer IDF versions didn't make sense before v1.20 was released but, now that it's out, there are a few of us that would like to spend time exploring 5.x...

**Comment by Jibun-no-Kage** [comment]
@mattytrentini , Hey, thanks for chiming in.  I was looking at the IDF integration to VS Code earlier today as well.  I have used VS Studio off and on for many years, but not used VS code as yet.  That said, I too need to spend some time with IDF, MP has let me avoid this.  I guess we should close this discussion for now, since MP support for C6 is  a bit out there, as @scruss noted, nothing definite as yet.  Once everything lines up, I for one will be more than happy to be using MP on C6.  Guess the next step for me is fire up my docker system.

**Reply by mattytrentini** [reply]
Docker _really_ helps when switching between toolchains! As a quick note, here's how I build for ESP32:

```bash
>docker run -ti --rm -v $(pwd):/code -w /code espressif/idf:v4.4.3 bash -c \
 "make -C mpy-cross && make -C ports/esp32 submodules all BOARD=GENERIC_S3_SPIRAM_OCT"
```

To try the 5.1 branch toolchain, just replace `v4.4.3` with `release-v5.1`:

```bash
>docker run -ti --rm -v $(pwd):/code -w /code espressif/idf:release-v5.1 bash -c \
 "make -C mpy-cross && make -C ports/esp32 submodules all BOARD=GENERIC_S3_SPIRAM_OCT"
```

Do remember to clean between builds:

```bash
>docker run -ti --rm -v $(pwd):/code -w /code espressif/idf:v4.4.3 bash -c \
 "make -C mpy-cross clean && make -C ports/esp32 clean BOARD=GENERIC_S3_SPIRAM_OCT"
```

Oh, and I mentioned that the change in toolchain causes build issues - which it does - but there are also breaking changes in the IDF that need to be updated. I was just reminded that the first issue is that the `esp_adc_cal` appears to have been moved...

**Comment by vencax** [comment]
I am interested as well. The ESP32-H2 support I guess is the same story.

**Comment by mattytrentini** [comment]
Just for a quick update...

All MicroPython ESP32 builds are now built with IDF v5 (specifically, currently, v5.0.2). Details can be found in #11528. 

The ESP32-C6, however, won't be supported until IDF v5.1 which is still in development. The development branch for [IDF v5.1](https://github.com/espressif/esp-idf/tree/release/v5.1) has gone through _two_ release candidates though, which - if history is a guide - means it's likely to be getting close to an official release...

I'm not sure about the status of IDF support for for the H2 @vencax, I'll try to find out more.

**Comment by Jibun-no-Kage** [comment]
Thanks for the update.  Awaiting IDF 5.1 with great interest, so MP for C6 can result.

**Comment by dirkonly** [comment]
Any progress on this? https://github.com/espressif/esp-idf/releases/tag/v5.1.1

**Comment by andrewleech** [comment]
I've started work on C6 support in #11869 but have only been working on it sporadically.

With IDF 5.0 they deprecated and changed a lot of APl interfaces, which is why 5.0 support took a while.

Esp generally provided compatibility layers though, to make it easier to port older software. Micropython is mostly using these to minimise the code change to get 5.0 support working.

On C6 however, being new, does not provide these compat layers, so it'll force a lot more re-write of the interfaces to support the newer peripheral api's.

I'm not sure if/when I'll get a chance to look at any more of these as every time I go near esp chips I get frustrated with the instability and painful cmake stuff pretty quickly!

**Reply by dustinlacewell** [reply]
Any luck?

**Comment by Jibun-no-Kage** [comment]
Yes, been waiting for some time now, surely things are moving in the right direction at this point?

**Reply by andrewleech** [reply]
Yes #11869 builds and seems to work now, bit of cleanup needed before I could hope to get it merged though.

**Comment by Jibun-no-Kage** [comment]
Just checking on status?  Is it merged yet?  Very much interested in using MP on C6!

**Comment by mattytrentini** [comment]
#11869 is in reasonable shape now. I've been testing it on the new [M5Stack NanoC6](https://shop.m5stack.com/products/m5stack-nanoc6-dev-kit) and the build appears functional and stable. It needs a review from core maintainers which will probably not occur until after v1.23 is released. Best to follow that PR for updates.

**Comment by Jibun-no-Kage** [comment]
We have any firm timeline at this point?  I see version 1.23 released, but no reference to ESP32-C6 on the official web download page?  Is it there and I missed it?

**Reply by andrewleech** [reply]
The MR for C6 support hasn't been merged yet, it's only been considered "ready for review" for a couple of days now since the last few problems have been addressed.

It has been slated for inclusion in the 1.24 release, however there's no fixed date for that yet.

I recommend just watching https://github.com/micropython/micropython/pull/11869 once it's finished and merged there should be some nightly builds available on the downloads page I think.

**Comment by Jibun-no-Kage** [comment]
So is there a rel;ease image at this point?

**Reply by mattytrentini** [reply]
It's still slated for inclusion in [v1.24](https://github.com/micropython/micropython/milestone/7) but that release has been delayed due to a number of complications (one of the more public issues has been the concerns with Errata 9 of the RP2350). I don't speak for the core team but I expect it to be merged soon (days, not months, hopefully not weeks).

**Comment by Jibun-no-Kage** [comment]
Yes, hope it is soon.  Be looking forward to it for some time now.

**Comment by vshymanskyy** [comment]
Does anyone know why ESP32-C6 build is so much bigger than ESP32-C3?
i.e. `v1.24.0-preview.447.g838f21298` (this is only `.app-bin`):
```
esp32-c3: 1673.6 KiB
esp32-c6: 1848.7 KiB
```

**Reply by mattytrentini** [reply]
I don't  _know_, but the C6 does have some additional features (most notably the low power core). I'd hazard a guess at the IDF _building more stuff_.

**Comment by mattytrentini** [comment]
ESP32-C6 support (#11869) was merged October 9th. 🎉
